### ISSUE TITLE
Problem with backward hook function

### ISSUE THREAD

**Issue by ludc** [pytorch/pytorch#598]
Hi,

there is something strange in the `backward` step (or maybe something I don't understand). If I define a Module that takes 3 inputs, the `grad_input` has to be of size 3, right ? But this is not the case here (from the backward_hook point of view):

```import torch
from torch.autograd import Variable
import torch.nn as nn
import torch.nn.functional as F

def bh(m,go,gi):
    print("Grad Input")
    print(go)
    print("Grad Output")
    print(gi)

class M(nn.Module):
    def __init__(self):
        super(M,self).__init__()
        self.register_backward_hook(bh)


    def forward(self,x,y,z):
        return (x+y+z)

x=Variable(torch.randn(1,5),requires_grad=True)
y=Variable(torch.randn(1,5),requires_grad=True)
z=Variable(torch.randn(1,5),requires_grad=True)

criterion=nn.MSELoss()
mod=M()
out=mod(x,y,z)
loss=criterion(out,Variable(torch.randn(1,5)))
loss.backward()```

In that case, when I print grad_input throught the hook function, it is just composed of two elements... Could you tell me where am I wrong ? But `x.grad, y.grad and z.grad` seem correctly computed

cc @ezyang @gchanan @zou3519 @SsnL @albanD @gqchen

**Comment by apaszke** [comment]
Ok, so the problem is that module hooks are actually registered on the last function that the module has created. In your case `x + y + z` is computed as `((x + y) + z)` so the hook is registered on that `(_ + z)` operation, and this is why you're getting only two grad inputs.

We'll definitely have to resolve this but it will need a large change in the autograd internals. However, right now @colesbury is rewriting them to make it possible to have multiple functions dispatched in parallel, and they would heavily conflict with his work. For now use only Variable hooks (or module hooks, but not on containers). Sorry!

**Comment by karandwivedi42** [comment]
@apaszke Is the refactor you referred to this one: https://github.com/pytorch/pytorch/pull/1016? Since that is merged, is this bug solvable now?

**Comment by apaszke** [comment]
Yes. We can remove the on hold label, but I don't have any good solutions in mind

**Comment by shubhamjain0594** [comment]
@apaszke Any progress or ideas on this issue, about how to solve. I may write a PR if required.

**Comment by apaszke** [comment]
No progress on this. It's actually a pretty hard problem to solve properly. I think I'd be ok with disallowing registering hooks on modules that have submodules as a partial workaround.

**Comment by whr94621** [comment]
@apaszke  Any progress or ideas on this issue? I also encounter this problem we the output of forward function is the result of ```torch.chunk```.

**Comment by apaszke** [comment]
Not yet. It's really hard to solve, and not such a common pitfall. I think we might just disable the function for now.

**Comment by t-vi** [comment]
So here is a copypaste from the duplicate I filed today.

The [pytorch documentation for `nn.Module.register_backward_hook`](http://pytorch.org/docs/master/nn.html#torch.nn.Module.register_backward_hook) says:

> The grad_input and grad_output may be tuples if the module has multiple inputs or outputs. The hook should not modify its arguments, but it can optionally return a new gradient with respect to input that will be used in place of grad_input in subsequent computation.

This is not quite accurate:
- `grad_output` is indeed the gradient of the loss w.r.t. the layer output. So if you have a layer l and do, say, `y = l(x) ; loss = y.sum(); loss.backward()`, you get the gradient of `loss` w.r.t. `y`. So far so good.
-  What is unexpected to most users is that `grad_input` are the inputs to the last operation in the layer. For linear layers, this is fairly complete, as the last op is `torch.addmm` multiplying the input with the weight and adding the bias. For other layers (e.g. do a Sequential, it’ll be the last op of the last layer, the inputs not even remotely related to the sequential layer’s inputs). You can see what will be used by looking at `y.grad_fn`.

This may be confusing to users, e.g. https://discuss.pytorch.org/t/exact-meaning-of-grad-input-and-grad-output/14186

So improving the docs is one thing, but should we make the hook actually live up to the description?

I see three strategies:
- Documentation only.
- The straightforward way of providing input gradients: collect the grad_ins with variable hooks and call the module hook when we have all of them. We loose the ability to return a different gradient.
- The somewhat convoluted way: If the module has hooks, wrap the module forward in a autograd function - similar to checkpointing. The the variable hook for the output would do the right thing.

**Comment by apaszke** [comment]
I think we should just remove this functionality for now (raise an error when `register_backward_hook` is called). It never worked correctly and is really hard to implement. I don't think the checkpointing way is good in this case, because it can cause certain backward hooks to trigger multiple times, and that's generally not what you want. It also doesn't work with `.grad`.

**Comment by karandwivedi42** [comment]
@apaszke Remove this functionality from all modules or only from Container type modules? If entirely removed, how should we obtain, for example, gradient norm?

**Comment by marcoancona** [comment]
I encountered the same issue. Is there any alternative to replace the gradient of a *sequence* of operations? Similarly to [custom_gradient](https://www.tensorflow.org/api_docs/python/tf/custom_gradient) in Tensorflow 1.8?

**Comment by tete1030** [comment]
Could this problem be properly noted in the doc before some solution?  Since it has existed so long and misleads many people including me.

**Comment by zou3519** [comment]
> Could this problem be properly noted in the doc before some solution? Since it has existed so long and misleads many people including me.

Sure, let's do that for now.

**Comment by cpuhrsch** [comment]
The corresponding pull requested https://github.com/pytorch/pytorch/pull/1016 mentioned by @apaszke has been marked as merged.

@ludc - is this issue resolved?

**Comment by apaszke** [comment]
No, this has never been fixed, and I think it's a wontfix for the foreseeable future.

**Comment by t-vi** [comment]
I must say I did like @albanD 's approach in #12573 . And think it might be worthwhile even if it has limitations (as long as hooks not being called is the only harm), one might warn if by the end of the forward the inputs have been inplace-modified..

Of course, my ideal PyTorch has inplace hooks for #7313, which might help with the issues that killed #12573.

**Comment by albanD** [comment]
@t-vi This approach lead to other issues with inplace ops in the autograd.
I will have time in the next couple of month to work on them properly and hopefully fix them all.

A note has been added in the documentation though explaining that they are mostly broken at the moment.

**Comment by gchanan** [comment]
added @albanD to assignee list based on the above conversation.

**Comment by yaroslavvb** [comment]
Can someone summarize why this turned out to be hard to implement?

**Comment by albanD** [comment]
Hi,

As many things touching the autograd engine, this is hard because of inplace operations!
We need a way to attach to all the differentiable inputs/outputs and get all the gradients for the input at once.

The current approach simply hooks onto the last Node associated with the first output. So for complex modules, it misses all intermediary results and is fairly broken.

The new approach I tried to introduce above was to add a specific Node into the graph that will be inserted at the input and output and will be able to call the callback with the saved data.
The current blocking (but that will be resolved soon 🤞 ) problem is that the autograd engine does not really allow for these NoOp Nodes. It also does not allow to return views of the input if there is more than one output (because in the case where an inplace operation is performed on this output, the graph need to be rewritten and we can only do that if a Node has a single output).

Not sure if it's really clear. Does that help?

**Comment by yaroslavvb** [comment]
Yes, makes sense. Restricting to single-output modules seems fine since that's the most common use-case.

PS, the motivation for module backward hooks is that the user code becomes simpler. Example -- implement https://arxiv.org/abs/1510.01799 . Currently this requires a module forward hook to save the activations, then backward hook to combine saved activations with backprops. Module backward hook would lets you do this in a single function

**Comment by AlexBinder** [comment]
Hi all,

is there any fix intended other than forbidding? I ask because there are people who implement "non-pure-gradient" backward attributions (e.g. LRP) via custom stuff in backward hooks in the context of explaining ML predictions. For all those people it would be helpful, while forbidding backward hooks would cause a bigger problem as far as I can see.


I think I stumbled over something related when implementing a backward attribution and using it on GPU:
nn.Conv2d with bias=True

in short: on the CPU i see the grad inputs which i would expect (input feature map x, conv-weights, conv-bias)
on the GPU i see 2 input shapes to what looks like a bias addition. 

code attached below. apologies to code style from public restrooms. in this case it creates an inconsistency CPU-GPU

*******
from __future__ import print_function, division
import torch
import torch.nn as nn
import numpy as np

def add_ahook(model):
    for module in model.modules():
        module.register_backward_hook(testbackwardhook)

def testbackwardhook(module, relevance_input, relevance_output):

    print( 'types of rel inp, relout', type(relevance_input), type(relevance_output), len(relevance_input), len(relevance_output))

    print( 'out[0].shape', relevance_output[0].shape)
    for k in range(len(relevance_input)):
      if hasattr(relevance_input[k],'shape'):
        print( k, 'relevance_input['+str(k)+'].shape', relevance_input[k].shape)
      else:
        print( k, 'relevance_input['+str(k)+']', relevance_input[k])

def runstuff():
  c=nn.Conv2d(in_channels=5, out_channels=9, kernel_size=3, stride=1,
                     padding=1, groups=1, bias=True, dilation=1)

  add_ahook(c)

  x=torch.ones((1,5,7,7))
  x.requires_grad=True  

  device2=torch.device('cuda:0')
  #device2=torch.device('cpu')

  x=x.to(device2)
  c=c.to(device2)

  z=c(x)
  print('forwardoutput',z.shape)
  torch.sum(z).backward(retain_graph=True)  

if __name__=='__main__':
  runstuff()
